### PR TITLE
Add Shared Scope Support and Support for Multiple Scopes Per Resource

### PR DESCRIPTION
**Pull request by dushaniw**
Fix wso2/product-apim#7292

- [x] Integration tests passed

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/APIManager.java [hunk 1]
```diff
@@ -453,9 +453,34 @@ String addResourceFile(Identifier identifier, String resourcePath, ResourceFile
      * @param tenantid   tenant id
      * @return true if the scope key is already available
      * @throws APIManagementException if failed to check the context availability
+     * @deprecated Use isScopeKeyAssignedLocally() instead.
      */
+    @Deprecated
     boolean isScopeKeyAssigned(APIIdentifier identifier, String scopeKey, int tenantid) throws APIManagementException;
```
Comment (tharindu1st) [inline components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/APIManager.java:459]: can we remove it

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/AMDefaultKeyManagerImpl.java [hunk 2]
```diff
@@ -706,14 +762,553 @@ public AccessTokenInfo getAccessTokenByConsumerKey(String consumerKey) throws AP
...
+    /**
+     * This method will be used to retrieve details of a Scope in the authorization server.
+     *
+     * @param scopeName    Scope Name to retrieve
+     * @param tenantDomain tenant domain to retrieve scope from
+     * @return Scope object
+     * @throws APIManagementException if an error while retrieving scope
+     */
+    @Override
+    public Scope getScopeByName(String scopeName, String tenantDomain) throws APIManagementException {
```
Comment (tharindu1st) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/AMDefaultKeyManagerImpl.java:921]: scopeName -> name

File: components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/APIManager.java (hunk 1 above)
Comment (dushaniw) [reply components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/APIManager.java:459]: This was added back as it seemed not a good practice to remove an API https://github.com/wso2/carbon-apimgt/pull/8409#discussion_r409378751

Comment (dushaniw) [reply components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/APIManager.java]: As the method is not used by any implementation other than internal carbon-apimgt methods, removed the method from interface. https://github.com/wso2/carbon-apimgt/pull/8476/commits/99c8da528391eede8b3d4adfab83a003c5c4ce7b

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/AMDefaultKeyManagerImpl.java (hunk 2 above)
Comment (dushaniw) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/AMDefaultKeyManagerImpl.java]: fixed https://github.com/wso2/carbon-apimgt/pull/8476/commits/c227ae14823523d58397d0190219fada595fd382

**Pull request merged**